dxfgerber.1
.\" -*- nroff -*-
.\" dxfgerber.1
.\" By: R.F. Smith <rsmith@xs4all.nl>
.\" $Date$
.\"
.Dd
.Os 
.Dt dxfgerber 1 CON
.Sh NAME
.Nm dxfgerber
.Nd Extract 
.Ql LINE 
and 
.Ql ARC 
entities from a Drawing eXchange Format 
.Pq Dq dxf
file and arranges them in a way suitable for optimized
cutting. 
.Sh SYNOPSIS
.Nm 
.Op Ar
.Sh DESCRIPTION
This manual page documents 
.Ql $Revision$
of the 
.Nm
program. 

The cutworks software that comes with a gerber cutter doesn't
automatically optimize the cutting paths it reads from dxf files. It
essentially cuts lines in the order it finds them in the dxf file.

This program reads a dxf file. The name of the file must end in 
.Ql .dxf
or
.Ql .DXF , 
otherwise the program will report an error and quit. It
extracts all the
.Ql LINE 
and 
.Ql ARC
entities from it. It then searches through all these entities and assembles
connected entities into lists called contours. If necessary, the direction of
entities in a contour is changed so that all entities can be cut in one
continuous movement.

These contours and any remaining lines and arcs are then sorted in ascending
order from the left-bottom corner of their bounding box.

It also converts arcs into line segments. By default these line segments are
no more than 5 mm in length. It ignores the
.Dv $MEASUREMENT
variable in the dxf file because that is often not set correctly and
assumes that the units in the dxf file are millimeters.

The output is written to a new file. The name of the new file is based on the
name of the old file, with "_mod" appended. So 
.Ql foo.dxf
becomes 
.Ql foo_mod.dxf .
.Sh EXIT STATUS
.Ex -std
.\".Sh DIAGNOSTICS
.Sh COMPATIBILITY
The 
.Nm
program requires the Python interpreter. It was written for version 2.7, but
should be able to work with 3.x after fixing with
.Xr 2to3 .
.Sh SEE ALSO
.Xr python 1 ,
.Xr 2to3.
.Rs
.%B Python Programming Language - Official Website
.%U http://www.python.org/
.Re
.Sh HISTORY
This program was written to stop having to fix up DXF files manually before
loading them into cutworks.
.Sh AUTHOR
This manual and the 
.Nm
software were written by 
.An Roland Smith Aq rsmith@xs4all.nl .
.Pp
The latest version of this program is available at:
.Lk http://rsmith.home.xs4all.nl/software/ 
.Sh LICENSE
To the extent possible under law, Roland Smith has waived all copyright and
related or neighboring rights to this manual. This work is published from the
Netherlands. See 
.Lk http://creativecommons.org/publicdomain/zero/1.0/
.Pp
The
.Nm
program itself is released under the two-clause BSD license given below;
.Pp
Copyright \(co 2011,2012 R.F. Smith <rsmith@xs4all.nl>. All rights reserved.
.Pp
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
.Bl -tag -width "foo" -compact
.It 1.
Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.
.It 2.
Redistributions in binary form must reproduce the above copyright notice, this
list of conditions and the following disclaimer in the documentation and/or
other materials provided with the distribution.
.El
.Pp
THIS SOFTWARE IS PROVIDED BY AUTHOR AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS
OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" EOF


dxfgerber.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Optimizes lines and arcs from a DXF file for cutting on a Gerber cutter,
# output another DXF file.
#
# Copyright © 2011,2012 R.F. Smith <rsmith@xs4all.nl>. All rights reserved.
# $Date$
# 
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY AUTHOR AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.

import sys
import datetime
import nctools.dxfgeom as dxfgeom
from nctools.fileutils import outname

__proginfo__ = ('dxfgerber [ver. ' + '$Revision$'[11:-2] +
                '] (' + '$Date$'[7:-2] + ')')


def dxf_header(progname, loc, loe):
    """Create the comments for at the beginning of a DXF file.

    :progname: name of the program
    :loc: list of Contours
    :loe: list of Entities
    :returns: a string containing the beginning of a DXF file
    """
    a = 'This conversion was started on {}'
    b = 'This file contains {} contours, {} remaining entities.'
    u = '#{} bounding box ({:.3f}, {:.3f}, {:.3f}, {:.3f})'
    dt = datetime.datetime.now()
    lines = ['999', 'DXF file generated by {}'.format(progname), '999',
            a.format(dt.strftime("%A, %B %d %H:%M")), '999', 
            b.format(len(loc), len(loe))]
    for n, c in enumerate(loc):
        lines += ['999', u.format(n+1, *c.getbb())]
    lines.append('')
    return '\n'.join(lines)


def start_entities():
    """Create the beginning of an entities section.

    :returns: a string to begin the ENTITIES section
    """
    return "  0\nSECTION\n  2\nENTITIES\n"


def end_entities():
    """Create the end of an entities section.
    
    :returns: a string to end the ENTITIES section
     """
    s = "  0\nENDSEC\n  0\nEOF"
    return s


def main(argv):
    """Main program for the dxfgerber utility.

    :argv: command line arguments
    """
    if len(argv) == 1:
        print __proginfo__
        print "Usage: {} [file.dxf ...]".format(argv[0])
        exit(1)
    del argv[0]
    for f in argv:
        if not f.endswith(('.dxf', '.DXF')):
            h = 'Probably not a DXF file. Skipping file "{}".'
            print h.format(f)
        try:
            ofn = outname(f, extension='.dxf', addenum='_mod')
            # Find entities
            entities = dxfgeom.fromfile(f)
        except ValueError as e:
            print e
            h = "Cannot construct output filename. Skipping file '{}'."
            print h.format(f)
            print "A valid filename _must_ have a '.dxf' extension."
            print "And it must be more than just the extension."
            continue
        except IOError as e:
            print "Cannot read file: {}".format(e)
            print "Skipping file '{}'".format(f)
            continue
        # Find contours
        (contours, rement) = dxfgeom.find_contours(entities)
        # Sort in y1, then in x1.
        entities = contours + rement
        entities.sort()
        # Output
        outf = open(ofn, 'w')
        outf.write(dxf_header(__proginfo__, contours, rement))
        outf.write(start_entities())
        for e in entities:
            outf.write(e.dxfdata())
        outf.write(end_entities())
        outf.close()


if __name__ == '__main__':
    main(sys.argv)
